clippy_lints/src/attrs.rs

```diff
@@ -518,6 +558,42 @@ impl<'tcx> LateLintPass<'tcx> for Attributes {
                     }
                 }
             },
+            ItemKind::Const(_, body_id) => {
+                // Looking for #[should_panic] attributes without a specified message means
+                // looking for the expanded #[test] const item.
+                //
+                // #[test] attributes expand to a const item that initializes `test::TestDescAndFn`.
+                //
+                // `#[test] #[should_panic] fn f() {}`
+                // becomes roughly:
+                // ```
+                // const f: test::TestDescAndFn = test::TestDescAndFn {
+                //   ...,
+                //   testfn: test::StaticTestFn(|| test::assert_result(f()))
+                // }
+                // ```
+                // The #[should_panic] attribute, along with its span, is on the `f`.
+
+                let body = cx.tcx.hir().body(body_id);
+
+                if let ExprKind::Struct(QPath::Resolved(_, path), fields, _) = body.value.kind
+                    && let Some(did) = path.res.opt_def_id()
+                    && match_def_path(cx, did, &paths::TEST_DESC_AND_FN)
+                    && let Some(testfn_field) = fields.iter().find(|field| field.ident.name == sym!(testfn))
+                    // test::StaticTestFn(|| ..)
+                    && let ExprKind::Call(_, [closure]) = testfn_field.expr.kind
+                    && let ExprKind::Closure(closure) = closure.kind
+                    // test::assert_result(f())
+                    //                     ^ get this `DefId` and its should_panic attr
+                    && let ExprKind::Call(_, [fn_call_closure]) = cx.tcx.hir().body(closure.body).value.kind
+                    && let ExprKind::Call(callee, _) = fn_call_closure.kind
+                    && let ExprKind::Path(QPath::Resolved(_, path)) = callee.kind
+                    && let Some(test_function_did) = path.res.opt_def_id()
+                    && let Some(attribute) = cx.tcx.get_attr(test_function_did, sym::should_panic)
```

Comment:
If the should_panic attribute remains on the function is there a reason we aren't checking that directly? Rather than finding the function via its test description

Reply:
Huh, okay I think I made a mistake when I rewrote this from pre expansion to late lint. It seems like I was wrong: `check_attribute` *does* get called for `#[should_panic]`. It must have been that I forgot to run with `--test`, in which case the function is completely cfg'd out and does not show up at all (so it makes sense that the attribute didn't get visited).
In that case we can make this a whole lot simpler.